### DISCUSSION TITLE
Структуры данных в C# - Лист, Стек, Очередь

### DISCUSSION THREAD

**Discussion by pingvin1308** [cleannetcode/CsharpForBeginers#13] [Новички]
Домашка: улучшите реализацию листа, стека и очереди добавив возможности работать с любыми типами данных. К примеру чтобы можно было сделать лист строк или лист классов Person :)
На данном этапе просьба не использовать дженерики (если вы знаете что это). По этому ключем к решению задачи будет object. Наша цель по итогу разобраться как устроены структуры данных + посмотреть чем же плоха будет наша реализация в сравнении с дженерик реализацией :)

Плюсом сделайте реализацию удаления элемента из листа и поправьте баг со стеком и очередью. Баг заключается в том, что когда все элементы удаляются из коллекции (Pull или Dequeue), в коллекции остается хвост или голова)

**Comment by coder1coder** [comment]
**Домашка** по теме **Лист, Стек, Очередь**
[https://github.com/coder1coder/Example](https://github.com/coder1coder/Example)

**Reply by pingvin1308** [reply]
Отлично! :)

**Reply by emedvedeva** [reply]
> **Домашка** по теме **Лист, Стек, Очередь**
> https://github.com/coder1coder/Example

Еще раз спасибо за комментарий моего домашнего задания. Позволь и мне оставить свой комментарий.
Очень понравился твой код, тоже хотела сначала сделать подобную реализацию удаления элемента из листа, но не смогла понять как сравнивать элементы типа Object.
И вопрос, а зачем ты разделил домашнее задание  на проекты.

**Reply by emedvedeva** [reply]
> > **Домашка** по теме **Лист, Стек, Очередь**
> > https://github.com/coder1coder/Example
> 
> Еще раз спасибо за комментарий моего домашнего задания. Позволь и мне оставить свой комментарий.
> Очень понравился твой код, тоже хотела сначала сделать подобную реализацию удаления элемента из листа, но не смогла понять как сравнивать элементы типа Object (про существование гугла я позабыла).
> И вопрос, а зачем ты разделил домашнее задание на проекты.

**Reply by pingvin1308** [reply]
Отчасти отвечу, просто он знает как связывать проекты. На данный момент для нас это не особенно важно)
У нас будет еще занятие про это, так что в будущем мы тоже так будем делать :)

**Reply by coder1coder** [reply]
> 
> 
> > **Домашка** по теме **Лист, Стек, Очередь**
> > https://github.com/coder1coder/Example
> 
> Еще раз спасибо за комментарий моего домашнего задания. Позволь и мне оставить свой комментарий.
> Очень понравился твой код, тоже хотела сначала сделать подобную реализацию удаления элемента из листа, но не смогла понять как сравнивать элементы типа Object.
> И вопрос, а зачем ты разделил домашнее задание на проекты.

Проект [Collections](https://github.com/coder1coder/Example/tree/main/Collections) в решении у меня это библиотека классов, которую я использую во втором [проекте-примере.](https://github.com/coder1coder/Example/tree/main/UseCustomCollection). Сделал для удобства дальнейшего использования в других проектах. Более правильно и своевременно вам на эти вопросы ответит Роман.

**Comment by emedvedeva** [comment]
> Домашка: улучшите реализацию листа, стека и очереди добавив возможности работать с любыми типами данных

Добрый день! Почему то в репозитории не нашла папку со вчерашним кодом или нужно самостоятельно с самого начала делать, а не продолжать?

**Reply by pingvin1308** [reply]
Добрый день! Нужно переключить ветку на develop :)
Не получается скинуть картинку, чтобы показать. На странице репозитория, есть кнопочка, на которой написано  Main и рядом стрелочка вниз)
Если нажать на эту кнопочку, что появится возможность выбрать ветку develop)

Правда в любом случаем при клонировании кода к себе на компьютер, придется переключиться на ветку develop в программе GitHub Desktop. Там есть похожая кнопка, но над ней написано Current branch (Текущая ветка)

**Comment by trulander** [comment]
Сделал домашку https://github.com/trulander/CsharpForBeginners/tree/develop/MyListImplementation
Реализовал удаление из листа. Удаляет текущий элемент списка и переключает указатель на следующий элемент, если следующего нет то переключаем на первый который в голове. Также доработал MoveNext таким образом чтобы при достижении конца списка выдавался false  как и было до этого, но переключаю указатель на первый элемент списка т.е на голову.

У stack и queue исправил ошибки при работе с последним элементом. как было указано в комментариях.

Переделал общий концепт хранимых данных, теперь можно хранить все что угодно. боксинг \ анбоксинг.

**Reply by pingvin1308** [reply]
Отлично сделал!)
Фигово, что я не пояснил как должно работать удаление из листа, но и так сойдет :)
Думаю на предстоящей домашке вы еще отработаете много навыков

**Comment by emedvedeva** [comment]
> Домашка: улучшите реализацию листа, стека и очереди добавив возможности работать с любыми типами данных. К примеру чтобы можно было сделать лист строк или лист классов Person :)
> На данном этапе просьба не использовать дженерики (если вы знаете что это). По этому ключем к решению задачи будет object. Наша цель по итогу разобраться как устроены структуры данных + посмотреть чем же плоха будет наша реализация в сравнении с дженерик реализацией :)
> 
> Плюсом сделайте реализацию удаления элемента из листа и поправьте баг со стеком и очередью. Баг заключается в том, что когда все элементы удаляются из коллекции (Pull или Dequeue), в коллекции остается хвост или голова)

У меня возник вопрос по поводу object, я не могу придумать как сделать так, чтобы потом можно было бы как то работать с элементами листа, то есть их нельзя будет делать с ними арифметические и вообще какие либо операции. Можно будет только добавлять и выводить в консоль. И вообще будет получаться лист с элементами разного типа. Или так и нужно сделать?

**Reply by trulander** [reply]
это с суботнего стрима, получается когда мы упаковываем в object то это боксинг, а анбоксинг нужно привести запакованные данные к нужному тебе типу по средством неявного преобразования прим (int)yourFieldOfTypeObject (string)yourFieldOfTypeObject
если в листе хранятся разные типы данных, то числа нужно проверять перед обработкой int.TryParse() ... Это со второго стрима, или с первого не помню точно.
Вообще хранить в листе разного типа данные это не про C#, так как здесь строгая типизация и нужно избегать реализации когда данные могут смешиваться, в качестве учебного примера думаю это нормально.

**Reply by trulander** [reply]
Посмотри в фигме https://www.figma.com/file/U7B9VO8IRr6cKjJuRotDvH/%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D1%8B-%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F-%D0%BD%D0%B0-C%23?node-id=0%3A1 Роман делал пометки, там есть пример с кодом где подписано что боксинг, а что анбоксинг

**Reply by emedvedeva** [reply]
> Вообще хранить в листе разного типа данные это не про C#, так как здесь строгая типизация и нужно избегать реализации когда данные могут смешиваться, в качестве учебного примера думаю это нормально.

вот я об этом же и говорю, если в листе будет object, то пользователь программы сможет запихать туда элементы разных типов. Как ты например сделал в своем листе: в списке и int и float  и string

**Reply by emedvedeva** [reply]
Если так то конечно, то получается, что нужно просто везде int поменять на object  и все. Если так, то домашнее задание в этот раз очень простое

**Reply by pingvin1308** [reply]
> Если так то конечно, то получается, что нужно просто везде int поменять на object и все. Если так, то домашнее задание в этот раз очень простое

Оно будет простым только для тех, кто понял суть изменения)
Такое часто бывает в программировании. Одно маленькое изменение решает много разных задач, но вот придумать как сделать такое маленькое изменение не всегда просто :)

Рад что домашка показалась простой ;)

**Comment by emedvedeva** [comment]
> Домашка: улучшите реализацию листа, стека и очереди добавив возможности работать с любыми типами данных. К примеру чтобы можно было сделать лист строк или лист классов Person :)
> На данном этапе просьба не использовать дженерики (если вы знаете что это). По этому ключем к решению задачи будет object. Наша цель по итогу разобраться как устроены структуры данных + посмотреть чем же плоха будет наша реализация в сравнении с дженерик реализацией :)
> 
> Плюсом сделайте реализацию удаления элемента из листа и поправьте баг со стеком и очередью. Баг заключается в том, что когда все элементы удаляются из коллекции (Pull или Dequeue), в коллекции остается хвост или голова)

Вот мое домашнее задание: https://github.com/emedvedeva/CsharpForBeginners/tree/main/MyListImplementation
1. изменила реализацию листа, добавив возможность работать с другими типами заменив тип значений в свойстве класса Box на Object
2 в стеке закомментировала поле _head, оно не нужно и теперь не остается
3. в очереди хвост занулила
4. реализовала удаление объекта из листа по индексу.

**Reply by pingvin1308** [reply]
Отличная работа!

**Reply by coder1coder** [reply]
Я тебе [тут комент](https://github.com/emedvedeva/CsharpForBeginners/commit/73909256cfd7c27581f3b1686857b2178be59bc7#r47027412) оставил, вдруг будет интересно мое мнение

**Comment by EvgenQ** [comment]
https://github.com/EvgenQ/List_Steck_Queue Домашка

**Reply by pingvin1308** [reply]
Привет, наконец дотянулся до проверок старых домашек. Как говорится лучше поздно чем никогда. Скинул тебе список задач на исправление в твоем репозитории

**Reply by pingvin1308** [reply]
В целом выглядит неплохо, нужно только сделать небольшие стилистические правки.
